tests/Unit/DependencyInjection/FOSHttpCacheExtensionTest.php

```diff
@@ -662,10 +665,62 @@ public function testVarnishCustomTagsHeader()
         $this->assertEquals(['tag_mode' => 'ban', 'tags_header' => 'myheader'], $container->getParameter('fos_http_cache.proxy_client.varnish.options'));
     }
 
+    /**
+     * @param string|array|mixed $serversValue           string and array are ok, anything else not, string only if it uses an env var
+     * @param string|mixed|null  $envValue               should be a string, otherwise an error will show up
+     * @param array|null         $expectedServersValue
+     * @param string|null        $expectExceptionClass
+     * @param string|null        $expectExceptionMessage
+     *
+     * @dataProvider dataVarnishServersConfig
+     */
+    public function testVarnishServersConfig($serversValue, $serversFromJsonEnvValue, $envValue, $expectedServersValue, $expectExceptionClass, $expectExceptionMessage)
+    {
+        if ($expectExceptionClass) {
+            $this->expectException($expectExceptionClass);
+            $this->expectExceptionMessage($expectExceptionMessage);
+        }
+
+        $_ENV['VARNISH_SERVERS'] = $envValue;
+        $container = $this->createContainer();
+
+        // workaround to get the possible env string into the EnvPlaceholderParamterBag
+        $container->setParameter('triggerServersValue', $serversValue);
+        $container->setParameter('triggerServersFromJsonEnvValue', $serversFromJsonEnvValue);
+        (new ResolveParameterPlaceHoldersPass())->process($container);
+
+        $config = $this->getBaseConfig();
+        $config['proxy_client']['varnish']['http']['servers'] = $container->getParameter('triggerServersValue');
+        $config['proxy_client']['varnish']['http']['servers_from_jsonenv'] = $container->getParameter('triggerServersFromJsonEnvValue');
+
+        $this->extension->load([$config], $container);
+        (new ResolveEnvPlaceholdersPass())->process($container);
+
+        $definition = $container->getDefinition('fos_http_cache.proxy_client.varnish.http_dispatcher');
```

Comment:
if we expect an exception, we should not get here. lets check for $expectExceptionClass before this again and return if we expected an exception. while trying stuff, i made the exception not be thrown and then got a confusing test failure on line 700.